### PR TITLE
Watchdog refurbishments

### PR DESCRIPTION
**Pull request by alexellis**
Watchdog refurbishments

Signed-off-by: Alex Ellis <alexellis2@gmail.com>

## Description

- Watchdog - allow new methods with and without body.
- Enforce hard-timeout via exec_timeout variable - gives *408 Request Timeout*
- Correct bug in timeouts for read/write of HTTP. via @stealthybox #98
- Documentation for new verbs and hard timeout.
- cgi_headers now defaults to true for functions.

## How Has This Been Tested?

- Gateway unaffected by commits

- Tested through new and existing unit tests

- Tested through watchdog binary outside of container:

```
exec_timeout=2 fprocess="node test.js" ./watchdog
```

test.js:

```
function t() {
  console.log("Stop");
}
console.log("Start");

setTimeout(t, 4000);
```

Result:

```
 
$ curl localhost:8080 -v
* Rebuilt URL to: localhost:8080/
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> GET / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.51.0
> Accept: */*
> 
< HTTP/1.1 408 Request Timeout
< Date: Mon, 03 Jul 2017 18:25:25 GMT
< Content-Length: 16
< Content-Type: text/plain; charset=utf-8
< 
Killed process.
* Curl_http_done: called premature == 0
* Connection #0 to host localhost left intact
```

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have signed-off my commits with `git commit -s`
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: watchdog/main.go [hunk 1]
```diff
@@ -115,24 +164,56 @@ func pipeRequest(config *WatchdogConfig, w http.ResponseWriter, r *http.Request)
...
-			pipeRequest(config, w, r)
-		} else {
+		switch r.Method {
+		case
+			"POST",
+			"PUT",
+			"DELETE",
+			"UPDATE":
+			pipeRequest(config, w, r, r.Method, true)
+			break
```
Comment (pmuens) [inline watchdog/main.go]: Why is a `break` added here?

File: watchdog/main.go [hunk 2]
```diff
@@ -115,24 +164,56 @@ func pipeRequest(config *WatchdogConfig, w http.ResponseWriter, r *http.Request)
...
+		case
+			"GET":
+			pipeRequest(config, w, r, r.Method, false)
+			break
```
Comment (pmuens) [inline watchdog/main.go]: Same question for this `break`

File: watchdog/requesthandler_test.go [hunk 3]
```diff
@@ -120,47 +121,102 @@ func TestHandler_HasXDurationSecondsHeader(t *testing.T) {
...
 	rr := httptest.NewRecorder()
 
 	req, err := http.NewRequest("GET", "/", nil)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	config := WatchdogConfig{}
+	config := WatchdogConfig{
+		// writeDebug:  true,
```
Comment (pmuens) [inline watchdog/requesthandler_test.go]: Should this comment be removed?

**Comment by alexellis** [comment]
Quick note on the GET method - `cat` will always fail for a `GET` - try changing it over to something that does not expect input such as `date` or `env`.

File: watchdog/config_test.go [hunk 4]
```diff
@@ -25,6 +25,32 @@ func (e EnvBucket) Getenv(key string) string {
 func (e EnvBucket) Setenv(key string, value string) {
 	e.Items[key] = value
 }
+
+func TestRead_CgiHeaders_OverideFalse(t *testing.T) {
```
Comment (pmuens) [inline watchdog/config_test.go]: `OverrideFalse`

File: watchdog/main.go [hunk 5]
```diff
@@ -41,65 +44,111 @@ func debugHeaders(source *http.Header, direction string) {
...
+		}()
+	}
+
+	// Only write body if this is appropriate for the method.
+	if hasBody {
+		// Write to pipe in separate go-routine to prevent blocking
+		go func() {
+			defer wg.Done()
+			writer.Write(requestBody)
+			writer.Close()
```
Comment (pmuens) [inline watchdog/main.go]: What about moving this to the top of the function and adding a `defer` here?

File: watchdog/main.go (hunk 1 above)
Comment (alexellis) [inline watchdog/main.go]: Golint was OK with it. It's more explicit about the flow.

File: watchdog/requesthandler_test.go (hunk 3 above)
Comment (alexellis) [inline watchdog/requesthandler_test.go]: Yes it can be

File: watchdog/main.go (hunk 5 above)
Comment (alexellis) [inline watchdog/main.go]: I think (should check) that this can make the close happen sooner.

**Comment by alexellis** [comment]
I like those check marks. Thanks for your work and comments!

I would expect `cat` to block and then timeout. Is that what you saw?

File: watchdog/main.go (hunk 1 above)
Comment (pmuens) [inline watchdog/main.go]: Gotcha. 👍

**Comment by alexellis** [comment]
> 🤔 just tried it today by running the compiled fwatchdog with fprocess="/bin/cat". Unfortunately no timeout when doing a GET request (waiting here for 2 minutes).

Try with the `exec_timeout` env-var set.

**Pull request merged**